### PR TITLE
426 Add minimal abm example based on previous example

### LINKED ISSUE #426: Make minimal ABM Example
For demonstration purpose and a quick run of the simulation, we need a minimal ABM Example.

### PR DESCRIPTION
**Pull request by khoanguyen-dev**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [ ] Tests for new functionality has been added
* [ ] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [x] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [x] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [x] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [ ] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request.
* [ ] Coverage report for new code is acceptable. 

Closes #426

### CODE REVIEW

File: cpp/examples/minimal_abm.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    auto t0         = mio::abm::TimePoint(0);
+    auto tmax       = mio::abm::TimePoint(0) + mio::abm::days(30);
+    auto sim = mio::abm::Simulation(t0, std::move(world));
+    sim.advance(tmax);
+
+    // The results are saved in a table with 9 rows.
+    // The first row is t = time, the others correspond to the number of people with a certain infection state at this time:
+    // S = Susceptible, E = Exposed, C= Carrier, I= Infected, I_s = Infected_Severe,
+    // I_c = Infected_Critical, R_C = Recovered_Carrier, R_I = Recovered_Infected, D = Dead
```
Comment (mknaranja) [inline cpp/examples/minimal_abm.cpp:141]: Please adapt to new infection state names.

File: cpp/examples/minimal_abm.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    }
+    add_household_group_to_world(world, threePersonHousehold_group);
+
+    // Assumed percentage of infection state at the beginning of the simulation.
+    double exposed_pct = 0.01, infected_pct = 0.08, carrier_pct = 0.05, recovered_pct = 0.01;
+    // Assign an infection state to each person.
+    // The infection states are chosen randomly. They are distributed according to the probabilites set in the example.
+    auto persons = world.get_persons();
+    double susceptible          = 1 - exposed_pct - infected_pct - carrier_pct - recovered_pct;
+    std::vector<double> weights = {susceptible, exposed_pct, carrier_pct, infected_pct / 2,
```
Comment (mknaranja) [inline cpp/examples/minimal_abm.cpp:73]: This initialization is prone to errors, as it assumes the number and position of infection states to be fixed. There had been an error discovered in: https://github.com/DLR-SC/memilio/pull/288

We should use a future-safe approach by using the states enum.

File: cpp/examples/minimal_abm.cpp (hunk 1 above)
Comment (khoanguyen-dev) [reply cpp/examples/minimal_abm.cpp:141]: Is this related to issue #387 ?

Comment (mknaranja) [reply cpp/examples/minimal_abm.cpp:141]: Yes but I was wrong. It is not yet merged, David's still working on. So maybe we change this later.

File: cpp/examples/minimal_abm.cpp (hunk 2 above)
Comment (khoanguyen-dev) [reply cpp/examples/minimal_abm.cpp]: The agents are now assigned a random enum state.

File: cpp/examples/abm_minimal.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    // Set global infection parameters (similar to infection parameters in SECIR model) and initialize the world
+    mio::abm::GlobalInfectionParameters infection_params;
+
+    // Set same infection parameter for all age groups. For example, the incubation period is 4 days. 
+    infection_params.get<mio::abm::IncubationPeriod>() = 4.;
+    
+    // Create the world with infection parameters.
+    auto world = mio::abm::World(infection_params);
+
+    // There are 3 households for each hosehold group.
```
Comment (DavidKerkmann) [inline cpp/examples/abm_minimal.cpp:35]:
```suggestion
    // There are 3 households for each household group.
```

File: cpp/examples/abm_minimal.cpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    int n_households = 3;
+
+    // For more than 1 family households we need families. These are parents and children and randoms (which are distributed like the data we have for these households).
+    auto child = mio::abm::HouseholdMember(); // A child is 50/50% 0-4 or 5-14.
+    child.set_age_weight(mio::abm::AgeGroup::Age0to4, 1);
+    child.set_age_weight(mio::abm::AgeGroup::Age5to14, 1);
+
+    auto parent = mio::abm::HouseholdMember(); // A parent is 50/50% 15-34, 35-59 or 60-79.
```
Comment (DavidKerkmann) [inline cpp/examples/abm_minimal.cpp:43]:
```suggestion
    auto parent = mio::abm::HouseholdMember(); // A parent is 50/50% 15-34 or 35-59.
```

File: cpp/examples/abm_minimal.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            person.set_assigned_location(work);
+        }
+    }
+
+    // During the lockdown, social events are closed for 90% of people. 
+    auto t_lockdown = mio::abm::TimePoint(0) + mio::abm::days(10);
+    mio::abm::close_social_events(t_lockdown, 0.9, world.get_migration_parameters());
+
+    auto t0         = mio::abm::TimePoint(0);
+    auto tmax       = mio::abm::TimePoint(0) + mio::abm::days(30);
```
Comment (DavidKerkmann) [inline cpp/examples/abm_minimal.cpp:129]: Consider changing to 60 days here or 30 days in the test definition.

**Review by DavidKerkmann — COMMENTED** [review]
Thanks for making the example. I can see the use of most of the core functionality, which was the aim of the minimal example. Just some minor comments, otherwise this can be merged.

File: cpp/examples/abm_minimal.cpp (hunk 5 above)
Comment (khoanguyen-dev) [reply cpp/examples/abm_minimal.cpp:129]: I now change the testing period to 30 days.

File: cpp/examples/abm_minimal.cpp (hunk 3 above)
Comment (xsaschako) [reply cpp/examples/abm_minimal.cpp]: If this is copy and pasted from the abm, next time you can change that there too :-)

**Review by xsaschako — APPROVED** [review]
As David said, looks good to me too!

**Pull request merged**